### PR TITLE
feat(blog): Introduce blog post on transitioning to pixi

### PR DESCRIPTION
**Pull request by ericmjl**
- Add a new blog post about switching from mamba to pixi.
- Discuss pixi's benefits for managing CPU and GPU environments.
- Highlight pixi's integration with Docker and containerization.
- Explain how pixi simplifies PyPI publishing and testing.
- Share experiences and tips on adopting pixi in development.
- Provide insights on pixi's potential to streamline workflows.

/schedule 2024-08-16

### CODE REVIEW

**Comment by ericmjl** [comment]
🚀 Deployed on https://deploy-preview-102--ericmjl-github-io.netlify.app

File: content/blog/its-time-to-try-out-pixi/contents.lr [hunk 1]
```diff
@@ -0,0 +1,995 @@
...
+as a data scientist and software developer.
+
+## What is pixi?
+
+In my mind, pixi is a package management multi-tool.
+`pixi` plays many roles, but here are the big, salient ones:
+
+| Role                | Analogy                                                  |
+| ------------------- | -------------------------------------------------------- |
+| Package installer   | `conda` and `pip`                                        |
```
Comment (ruben-arts) [inline content/blog/its-time-to-try-out-pixi/contents.lr:23]: You could possibly also add `pipx` or `apt-get` and `homebrew` which is what we're trying to replace with `pixi global`

File: content/blog/its-time-to-try-out-pixi/contents.lr [hunk 2]
```diff
@@ -0,0 +1,995 @@
...
+echo 'eval "$(pixi completion --shell zsh)"' >> ~/.zshrc
+```
+
+### Starting a new project with `pixi` enabled
+
+For those who are used to `conda` idioms,
+`pixi` won't create a centralized `conda` environment
+but instead will create a `.pixi` sub-directory
+within the directory of your `pixi` configuration file
+(either `pyproject.toml` or `pixi.toml`).
```
Comment (ruben-arts) [inline content/blog/its-time-to-try-out-pixi/contents.lr:142]: You could possibly add that this is working like:
- `npm`:  Nodejs ecosystem
- `cargo`: Rust ecosystem
- `poetry`: another pypi tool
If that would help your readers understand the paradigm

File: content/blog/its-time-to-try-out-pixi/contents.lr [hunk 3]
```diff
@@ -0,0 +1,995 @@
...
+In line with my desire to be able to specify sets of dependencies
+for runtime environments and development environments
+and also minimize the number of configuration files present,
+it makes sense to initialize with a `project.toml` configuration
```
Comment (ruben-arts) [inline content/blog/its-time-to-try-out-pixi/contents.lr:146]:
```suggestion
it makes sense to initialize with a `pyproject.toml` configuration
```

File: content/blog/its-time-to-try-out-pixi/contents.lr [hunk 4]
```diff
@@ -0,0 +1,995 @@
...
+rather than the default `pixi.toml` configuration file.
+This is done by executing:
+
+```bash
+pixi init --pyproject -c conda-forge -v
```
Comment (ruben-arts) [inline content/blog/its-time-to-try-out-pixi/contents.lr:151]: To not get a deprecation warning in the latest release (of today) the new cli would be:
```suggestion
pixi init --format pyproject -c conda-forge -v
```
Also you could leave out `-c conda-forge` as we default to that, if you would like that.
```suggestion
pixi init --format pyproject  -v
```

File: content/blog/its-time-to-try-out-pixi/contents.lr [hunk 5]
```diff
@@ -0,0 +1,995 @@
...
+```
+
+This gives me the following `pyproject.toml` file:
+
+```toml
+[project]
+name = "pixi-cuda-environment" # defaults to my repo name.
```
Comment (ruben-arts) [inline content/blog/its-time-to-try-out-pixi/contents.lr:158]:
```suggestion
name = "pixi-cuda-environment" # defaults to my directory name.
```

File: content/blog/its-time-to-try-out-pixi/contents.lr [hunk 6]
```diff
@@ -0,0 +1,995 @@
...
+platforms = ["osx-arm64"]
+
+[tool.pixi.pypi-dependencies]
+pixi-cuda-environment = { path = ".", editable = true }
+
+[tool.pixi.tasks]
+```
+
+What's important to note here is that `[project] -> dependencies`
+lists the Python project's runtime dependencies,
```
Comment (ruben-arts) [inline content/blog/its-time-to-try-out-pixi/contents.lr:180]: pixi doesn't act differently on these dependencies than `pip` or `uv` would. It interprets them as `PyPI` dependencies. This is also the case for the `[project.optional-dependencies]` except that they're automatically put into the "feature" named by the optional key.

e.g. `pixi` sees these two configurations as equal:
```toml
[project]
dependencies = [
    "docutils",
    "BazSpam == 1.1",
]

[project.optional-dependencies]
PDF = ["ReportLab>=1.2", "RXP"]
```
```toml
[tool.pixi.pypi-dependencies]
docutils = "*"
BazSpam = "== 1.1"

[tool.pixi.feature.PDF.pypi-dependencies]
ReportLab = ">=1.2"
RXP = "*"
```

File: content/blog/its-time-to-try-out-pixi/contents.lr [hunk 7]
```diff
@@ -0,0 +1,995 @@
...
+
+If you went cold turkey as I did but still need to install Python tooling globally,
+then there's muscle memory that you'll need to un-learn very quickly.
+Whereas previously I could rely on a base environment
+through my `mambaforge` installation, it no longer exists.
+Rather, I needed to install Python and `pipx` globally with `pixi`
+and then use `pipx` (not `pixi`!) to install that tool using
+
+```bash
+pipx global install --python $(which python) <tool name>
```
Comment (ruben-arts) [inline content/blog/its-time-to-try-out-pixi/contents.lr:925]:
```suggestion
pipx install --python $(which python) <tool name>
```
`global` is a pixi thing :wink:

**Review by ruben-arts — COMMENTED** [review]
Hi @ericmjl, this is amazing! Such a cool view from a python user perspective. 

On request of Wolf, I did a review of the blog. And I love it!

I left a few comments to make sure you have the same understanding we do of pixi. Do with this what you want! 

I'm testing your docker example. I'm surprised by the nvidia-smi example. Will let you know if anything comes up!

Again thanks and looking forward to getting this published!

File: content/blog/its-time-to-try-out-pixi/contents.lr [hunk 8]
```diff
@@ -0,0 +1,995 @@
...
+ => [7/7] RUN chmod 700 /repo/entrypoint.sh                         0.3s
+ => exporting to image                                             22.2s
+ => => exporting layers                                            22.2s
+ => => writing image sha256:6f13e2a7b362c7b3736d4405f7f5566775320d  0.0s
+```
+
+To verify that I could indeed run CUDA-accelerated JAX, I entered into the container:
+
+```shell
+❯ docker run --gpus all -it docker.io/library/pixi-cuda /bin/bash
```
Comment (ruben-arts) [inline content/blog/its-time-to-try-out-pixi/contents.lr:506]: I didn't know this existed, but after following the following guide it did in deed work :100:  https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/latest/install-guide.html

**Review by ruben-arts — COMMENTED** [review]
Verified my concern and it works amazingly!

**Comment by juanitorduz** [comment]
Very cool post! It helped me with a project migration :)

File: content/blog/its-time-to-try-out-pixi/contents.lr (hunk 1 above)
Comment (ericmjl) [reply content/blog/its-time-to-try-out-pixi/contents.lr:23]: Yes, I think that's a great point! I have added it as a new line, "global tool installer", which feels categorically different from a package installer alone.

File: content/blog/its-time-to-try-out-pixi/contents.lr (hunk 2 above)
Comment (ericmjl) [reply content/blog/its-time-to-try-out-pixi/contents.lr:142]: That's a great point. I've just updated the text to allude to this point!

File: content/blog/its-time-to-try-out-pixi/contents.lr (hunk 6 above)
Comment (ericmjl) [reply content/blog/its-time-to-try-out-pixi/contents.lr:184]: Nice point! Thank you for educating me on this, @ruben-arts!

**Comment by ericmjl** [comment]
:hourglass: **Merge Schedule**
Scheduled to be merged on 2024-08-16 00:00:00 (UTC)

File: content/blog/its-time-to-try-out-pixi/contents.lr [hunk 9]
```diff
@@ -0,0 +1,1040 @@
...
+there will be a Jupyter kernel named `Pixi - Python 3 (ipykernel)`
+which will contain the packages contained in your environment.
+If you commit the notebooks + pixi configuration files to source control,
+someone else can download it and reproduce the environment easily.
+And as long as you don't rely on data living at a hard-coded path,
+that other person should be able to reproduce your work as well!
+
+## Acknowledgments
+
+I'd like to thank Sean Law and Reuben Arts for reviewing this content,
```
Comment (ruben-arts) [inline content/blog/its-time-to-try-out-pixi/contents.lr:1017]:
```suggestion
I'd like to thank Sean Law and Ruben Arts for reviewing this content,
```
I guess you meant me :wink:

**Comment by ruben-arts** [comment]
Btw, fun fact (or you where already aware), but this is scheduled to be released exactly one year after launch: https://prefix.dev/blog/launching_pixi

File: content/blog/its-time-to-try-out-pixi/contents.lr (hunk 9 above)
Comment (ericmjl) [reply content/blog/its-time-to-try-out-pixi/contents.lr]: Yes, that is right, and thank you for the catch, @ruben-arts!

**Comment by ericmjl** [comment]
> Btw, fun fact (or you where already aware), but this is scheduled to be released exactly one year after launch: https://prefix.dev/blog/launching_pixi

Wow, that is such a coincidence!

File: content/blog/its-time-to-try-out-pixi/contents.lr [hunk 10]
```diff
@@ -0,0 +1,1048 @@
...
+As such, I need to know how to use `pixi` to install GPU-enabled packages.
+I went to my go-to, JAX and PyTorch,
+to figure out how to install the GPU-enabled versions of these packages.
+
+To make this happen, we must distinguish between both environments:
+CPU-only, like on my Mac,
+and GPU-enabled, like on my home lab that runs Ubuntu.
+First, we define the common and CUDA-specific dependencies on CUDA-enabled systems.
+To start, my `pyproject.toml` file changes a bit on the `dependencies` section
+within the `pyproject.toml` file:
```
Comment (Armavica) [inline content/blog/its-time-to-try-out-pixi/contents.lr:331]: Did you mean to mention `pyproject.toml` twice in this sentence?

**Comment by ericmjl** [comment]
> Thank you for this useful blog post, Eric! I have questions:
> 
> * You mention the advantage of putting runtime dependencies in `project.dependencies`, so is there a reason why you put `jax` and `pytorch` in `tool.pixi.dependencies` later?

Good point. I think it was to simplify explanation.

> * Why is it `jax = "*"` and not with a pinned specification like the others dependencies? Was this not added with `pixi add`?

That's right! I just allowed a floating version to 

> * Will `pixi` also respect `project.dependencies` if it is setup with `pixi.toml` instead of `pyproject.toml`?

If it's set up with `pixi.toml`, I am not sure of the behaviour, but my guess it that it'll ignore `project.dependencies` as my hunch says it might not be part of what `pixi` looks for.

File: content/blog/its-time-to-try-out-pixi/contents.lr (hunk 10 above)
Comment (ericmjl) [reply content/blog/its-time-to-try-out-pixi/contents.lr:331]: Great catch, thank you!

**Pull request merged**